src/System.Collections/src/System/Collections/BitArray.cs

```diff
@@ -74,40 +77,40 @@ public BitArray(byte[] bytes)
             // this value is chosen to prevent overflow when computing m_length.
             // m_length is of type int32 and is exposed as a property, so 
             // type of m_length can't be changed to accommodate.
-            if (bytes.Length > int.MaxValue / BitsPerByte)
+            if (bytes.Length > int.MaxValue >> BitShiftPerByte) // Divide by 8.
             {
                 throw new ArgumentException(SR.Format(SR.Argument_ArrayTooLarge, BitsPerByte), nameof(bytes));
             }
 
-            m_array = new int[GetArrayLength(bytes.Length, BytesPerInt32)];
+            m_array = new int[GetIntegerArrayLengthFromByteLength(bytes.Length)];
             m_length = bytes.Length * BitsPerByte;
 
             int i = 0;
             int j = 0;
-            while (bytes.Length - j >= 4)
+            int totalCount = bytes.Length >> 2;     // Divide by 4.
```

Comment:
I'm surprised `bytes.Length >> 2` here produces better code than `bytes.Length / 4`, given that the JIT knows bytes.Length is non-negative.

Reply:
I didn't verify in WinDbg what the actual disassembly would look like. Maybe it's fine (and if so, will revert). Did a quick test on [sharplab](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAGABJgRgG4BYAKEwGZ8AmXAYVwG9zc39qBLAOzlwFkCACmABPOAFMA2gF1cYycgCUzVu3WYA7PPETkAOgAyE7gHM4AC1wA+a7hqky6gL5q2VXDz78aI3bJ1FFRYndXYtQL0jE3MrAHpcKEcXcmcgA==):
```C#
    public int M1(byte[] bytes) {
        return bytes.Length >> 2;
    }
    public int M2(byte[] bytes) {
        return bytes.Length / 4;
    }
```

```text
C.M1(Byte[])
    L0000: mov eax, [rdx+0x8]
    L0003: sar eax, 0x2
    L0006: ret

C.M2(Byte[])
    L0000: mov eax, [rdx+0x8]
    L0003: mov edx, eax
    L0005: sar edx, 0x1f
    L0008: and edx, 0x3
    L000b: add edx, eax
    L000d: mov eax, edx
    L000f: sar eax, 0x2
    L0012: ret
```

cc @AndyAyersMS

Reply:
``` ini

BenchmarkDotNet=v0.11.2, OS=macOS High Sierra 10.13.3 (17D47) [Darwin 17.4.0]
Intel Core i7-4870HQ CPU 2.50GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.1.300-preview2-008062
  [Host]     : .NET Core ? (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.0-preview2-26130-06 (CoreCLR 4.6.26130.05, CoreFX 4.6.26130.01), 64bit RyuJIT


```
|   Method |     Mean |     Error |    StdDev | Ratio |
|--------- |---------:|----------:|----------:|------:|
|      Div | 192.2 ns | 0.9134 ns | 0.8097 ns |  1.00 |
| BitShift | 151.3 ns | 2.9729 ns | 2.9198 ns |  0.79 |


```C#
    [Benchmark(Baseline = true)]
    public int Div()
    {
        int result = 0;
        for (int i = 0; i < N; i++)
        {
            result ^= data_2D[i].Length / 4;
        }
        return result;
    }

    [Benchmark]
    public int BitShift()
    {
        int result = 0;
        for (int i = 0; i < N; i++)
        {
            result ^= data_2D[i].Length >> 2;
        }
        return result;
    }
```

Reply:
And here is the disassembly: https://www.diffchecker.com/tf6IPg8N

18 vs 22 instructions

Reply:
Well, until that can be addressed by the JIT, I'd personally prefer to see `(uint)arr.Length / 4` rather than `arr.Length >> 2`, as I believe the former is clearer about the intent (and I think you do, too, since you added a comment highlighting what it really meant :))  Presumably adding the cast makes the JIT happy?

Reply:
> Presumably adding the cast makes the JIT happy?

Basically it makes CSC generate an unsigned division because the JIT doesn't do it.